### ISSUE TITLE
DNS fails for .local files when aiodns is also installed

### ISSUE THREAD

**Issue by ColinRobbins** [aio-libs/aiodns#127]
### Describe the bug

Home Assistant uses aiohttp with aiodns.
The HA Overkiz integration has stopped working in 2024.11.
The issue has been logged in HA https://github.com/home-assistant/core/issues/132318 and traced to the use of aiodns.

The library https://github.com/iMicknl/python-overkiz-api has a test script.   
The test works fine when aiodns is not installed, but fails when aiodsn is installed.

The URL that fails to resolve is https://gateway-0403-xxxx-7195.local:8443.

If I downgrade to aiohttp version 3.10.3 the DNS resolution work as expected.

### To Reproduce

Run the Overkiz integration in HA, with a local server.

### Expected behavior

The DNS name should resolve correctly.

### Logs/tracebacks

```python-traceback
See  https://github.com/home-assistant/core/issues/132318
```

### Python Version

```console
$ python --version
3.12

Issue also occurs with 3.13.
```

### aiohttp Version

```console
$ python -m pip show aiohttp
3.11.9
```

### multidict Version

```console
$ python -m pip show multidict
6.1.0
```

### propcache Version

```console
$ python -m pip show propcache
0.2.0
```

### yarl Version

```console
$ python -m pip show yarl
1.18.0
```

### OS

Ubunto Linux

### Related component

Server

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by ColinRobbins** [comment]
Here is the output of the test...

```
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ cat dns_test.py 
import aiodns
import asyncio
import socket

ADDRESS = "gateway-0403-XXXX-7195.local"

async def main():
    resolver = aiodns.DNSResolver()
    try:
        result = await resolver.query(ADDRESS, "A")
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.gethostbyname(ADDRESS, socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.getaddrinfo(ADDRESS, 0, socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)


asyncio.run(main())
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ python3.12 dns_test.py 
(11, 'Could not contact DNS servers')
(11, 'Could not contact DNS servers')
(11, 'Could not contact DNS servers')
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ pip3 show aiodns
Name: aiodns
Version: 3.2.0
Summary: Simple DNS resolver for asyncio
Home-page: https://github.com/saghul/aiodns
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /home/colin/github/Home-Assistant-Core/venv/lib/python3.12/site-packages
Requires: pycares
Required-by: homeassistantSkip to content
Navigation Menu
aio-libs
aiohttp

Type / to search
Code
Issues
191
Pull requests
58
Discussions
Actions
Security
14
Insights
DNS fails for .local files when aiodns is also installed #10111
Open
1 task done
ColinRobbins opened this issue yesterday · 27 comments
Open
1 task done
DNS fails for .local files when aiodns is also installed
#10111
ColinRobbins opened this issue yesterday · 27 comments
Comments
@ColinRobbins
ColinRobbins commented yesterday
Describe the bug
Home Assistant uses aiohttp with aiodns.
The HA Overkiz integration has stopped working in 2024.11.
The issue has been logged in HA home-assistant/core#132318 and traced to the use of aiodns.

The library https://github.com/iMicknl/python-overkiz-api has a test script.
The test works fine when aiodns is not installed, but fails when aiodsn is installed.

The URL that fails to resolve is https://gateway-0403-xxxx-7195.local:8443.

If I downgrade to aiohttp version 3.10.3 the DNS resolution work as expected.

To Reproduce
Run the Overkiz integration in HA, with a local server.

Expected behavior
The DNS name should resolve correctly.

Logs/tracebacks
See  https://github.com/home-assistant/core/issues/132318
Python Version
$ python --version
3.12

Issue also occurs with 3.13.
aiohttp Version
$ python -m pip show aiohttp
3.11.9
multidict Version
$ python -m pip show multidict
6.1.0
propcache Version
$ python -m pip show propcache
0.2.0
yarl Version
$ python -m pip show yarl
1.18.0
OS
Ubunto Linux

Related component
Server

Additional context
No response

Code of Conduct
 I agree to follow the aio-libs Code of Conduct
@ColinRobbins ColinRobbins added the bug label yesterday
@bdraco
Member
bdraco commented yesterday • 
HA has used aiodns way before 2024.11

I don't think thats the cause. Something else must have changed.

@bdraco bdraco added the needs-info label yesterday
@ColinRobbins
Author
ColinRobbins commented yesterday • 
But the evidence suggests it is a change in how aiohttp works with it in some way.
It all works fine with 3.10.3, but breaks with 3.11.9

@Dreamsorcerer
Member
Dreamsorcerer commented yesterday
I think we started using aiohappyeyeballs in 3.11, so more likely that's the change which makes the difference. Not sure how that plays with aiodns.

@ColinRobbins
Author
ColinRobbins commented yesterday
I'm not familiar with these packages.
Are there any local tests I can do try and narrow down the issue?

@bdraco
Member
bdraco commented yesterday
I think we started using aiohappyeyeballs in 3.11, so more likely that's the change which makes the difference. Not sure how that plays with aiodns.

aiohappyeyeballs was 3.10 so I don't think that is related

@bdraco
Member
bdraco commented yesterday • 
But the evidence suggests it is a change in how aiohttp works with it in some way. It all works fine with 3.10.3, but breaks with 3.11.9

While that's helpful to narrow the window, its not enough to go on to figure out whats going wrong.

We need some sample code (ie small reproducer) that is expected to work but now fails.

@bdraco bdraco added the reproducer: missing label yesterday
@ColinRobbins
Author
ColinRobbins commented 19 hours ago
Here is some code that will reproduce the error without the need for any local hardware.

mport asyncio
import time
import aiohttp

from pyoverkiz.client import OverkizClient
from pyoverkiz.const import SUPPORTED_SERVERS, OverkizServer
from pyoverkiz.enums import Server

LOCAL_GATEWAY = "gateway-0403-XXXX-7195.local" # or use the IP address of your gateway

async def main() -> None:
    token = "" # you can set the token here for testing purposes, to re-use an earlier generated token

    # Local Connection
    session = aiohttp.ClientSession(
        connector=aiohttp.TCPConnector(verify_ssl=True))

    async with OverkizClient(
        username="", password="", token=token, session=session, verify_ssl=True, server=OverkizServer(
            name="Somfy TaHoma (local)",
            endpoint=f"https://{LOCAL_GATEWAY}:8443/enduser-mobile-web/1/enduserAPI/",
            manufacturer="Somfy",
            configuration_url=None,
        )
    ) as client:
        await client.login()

asyncio.run(main())
If you run this WITH the aiodns package installed, you get the error...

iohttp.client_exceptions.ClientConnectorDNSError: Cannot connect to host gateway-0403-xxxx-7195.local:8443 ssl:<ssl.SSLContext object at 0x7961589a1b50> [Could not contact DNS servers]
which in an exact reproduction of the error in HA / pyoverkiz test code.

if you run this WITHOUT the aiodns package installed, you get the error...

aiohttp.client_exceptions.ClientConnectorDNSError: Cannot connect to host gateway-0403-xxxx-7195.local:8443 ssl:<ssl.SSLContext object at 0x71b65b1d88d0> [Name or service not known]
This is "good" in the test sense as the DNS call is attempting a lookup and failing.

@ColinRobbins ColinRobbins mentioned this issue 19 hours ago
Local API cannot connect ('Could not contact DNS servers') in HA >2024.12.0 in Overkiz home-assistant/core#132318
Open
@bdraco
Member
bdraco commented 19 hours ago • 
We need a minimal reproducer that doesn't use any external libraries, and doesn't require any special hardware to run so our team can reproduce the issue on their systems. It should only use aiohttp (or its listed dependencies)

@bdraco
Member
bdraco commented 17 hours ago • 
Also you mentioned downgrading to 3.10.3 solves the problem. Its unclear if you mean that 3.10.4 is where the problem was introduced. It would be helpful to know the first version of aiohttp that you can reproduce the problem with.

@ColinRobbins
Author
ColinRobbins commented 16 hours ago • 
Here is some code without the library dependency...

import time
import aiohttp

LOCAL_GATEWAY = "gateway-0403-5459-7195.local" 

async def main() -> None:

    # Local Connection
    session = aiohttp.ClientSession(
        connector=aiohttp.TCPConnector(verify_ssl=True))

    async with session.post(
        f"https://{LOCAL_GATEWAY}:8443/xx/",
        data="",
        headers="",
        ssl=False,
    ) as response:
        print ("Made it")

asyncio.run(main())
This generates the error...

Traceback (most recent call last):
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/resolver.py", line 103, in resolve
    resp = await self._resolver.getaddrinfo(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
aiodns.error.DNSError: (11, 'Could not contact DNS servers')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 1348, in _create_direct_connection
    hosts = await self._resolve_host(host, port, traces=traces)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 989, in _resolve_host
    return await asyncio.shield(resolved_host_task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 1020, in _resolve_host_with_throttle
    addrs = await self._resolver.resolve(host, port, family=self._family)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/resolver.py", line 112, in resolve
    raise OSError(None, msg) from exc
OSError: [Errno None] Could not contact DNS servers

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/colin/github/python-overkiz-api/aiotest.py", line 22, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/aiotest.py", line 14, in main
    async with session.post(
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/client.py", line 1423, in __aenter__
    self._resp: _RetType = await self._coro
                           ^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/client.py", line 701, in _request
    conn = await self._connector.connect(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 544, in connect
    proto = await self._create_connection(req, traces, timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 1052, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/colin/github/python-overkiz-api/venv/lib/python3.12/site-packages/aiohttp/connector.py", line 1354, in _create_direct_connection
    raise ClientConnectorDNSError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorDNSError: Cannot connect to host gateway-0403-5459-7195.local:8443 ssl:False [Could not contact DNS servers]
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x70fae729b650>

Under 3.10 the DNS all resolves fine, and it makes contact with the host
Under 3.11.9 without aiodns, it all works fine and makes contact with the host
Under 3.11.9 with aiodns I get the error above

In terms of test code, I sadly don't have a way I can find of simulating the server.

I tried running a noddy server on my local machine using ncat...

while true ; do ncat -v -l -p 8443 -c 'echo "HTTP/1.1 200 OK\n\n $(date)"'; done
and then change the test code to

LOCAL_GATEWAY = "xxx.local" 
then it works as expected.

So it seems to me that the call from aiodns to getaddrinfo is getting an unexpected response, and logging it as a DNS error. Its not clear if it is trying to contact the server or not.

I don't know enough about getaddinfo to progress any further.

@bdraco
Member
bdraco commented 16 hours ago
Under 3.10 the DNS all resolves fine, and it makes contact with the host

Would you please clarify if 3.10.10 works as expected?

I'm guessing the problem was introduced in 3.11.x based on your comments but I'm not clear if 3.10.3 is the only working version in 3.10.x or not.

@ColinRobbins
Author
ColinRobbins commented 16 hours ago
Good point. I've cycled through multiple combinations of aiohttp and aiodns verisons.

The aiohttp version does NOT seem to matter after all - I should have tried that earlier. Sorry

The error occurs when upgrading from aiodns 3.1.0 to 3.2.0, irrespective of the aiohttp version.

@ColinRobbins
Author
ColinRobbins commented 16 hours ago
I note aiodns 3.2 added support for getaddrinfo, and its getaddrinfo that is failing.
I'll raise an issue in aiodsn.

@bdraco
Member
bdraco commented 16 hours ago
There isn't anything significant changed that would affect resolution between 3.1.0 and 3.2.0 aio-libs/aiodns@v3.1.0...v3.2.0 as it mostly adds some new methods.

Does the installed version of pycares or c-areas affect the ability to reproduce the issue?

@ColinRobbins ColinRobbins mentioned this issue 16 hours ago
getaddrinfo is failing with a local server in version 3.2 aio-libs/aiodns#126
Open
@bdraco
Member
bdraco commented 16 hours ago • 
I just tried the following and was able to get an mdns .local name to resolve in a Home Assistant container with aiodns

All three methods, .query, .gethostbyname, and .getaddrinfo are able to resolve the name.

homeassistant:/config/issue_10111# cat dns_test.py 
import aiodns
import asyncio
import socket


async def main():
    resolver = aiodns.DNSResolver()
    try:
        result = await resolver.query("drivewaygate.local.", "A")
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.gethostbyname("drivewaygate.local.", socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.getaddrinfo("drivewaygate.local.", 0, socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)


asyncio.run(main())
homeassistant:/config/issue_10111# python3 dns_test.py 
[<ares_query_a_result> host=192.168.212.44, ttl=60]
<ares_host_result> name=drivewaygate.local, aliases=[], addresses=['192.168.212.44']
<ares_addrinfo_result> cnames=[], nodes=[<ares_addrinfo_node_result> ttl=60, flags=0, family=2, socktype=0, protocol=0, addr=(b'192.168.212.44', 2)]
homeassistant:/config/issue_10111# pip3 show aiodns
Name: aiodns
Version: 3.2.0
Summary: Simple DNS resolver for asyncio
Home-page: https://github.com/saghul/aiodns
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /usr/local/lib/python3.13/site-packages
Requires: pycares
Required-by: aiodiscover, easyenergy, forecast-solar, homeassistant, pyjvcprojector, radios, slixmpp, vulcan-api
homeassistant:/config/issue_10111# pip3 show pycares
Name: pycares
Version: 4.5.0
Summary: Python interface for c-ares
Home-page: http://github.com/saghul/pycares
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /usr/local/lib/python3.13/site-packages
Requires: cffi
Required-by: aiodns
@ColinRobbins
Author
ColinRobbins commented 16 hours ago
There isn't anything significant changed that would affect resolution between 3.1.0 and 3.2.0 aio-libs/aiodns@v3.1.0...v3.2.0 as it mostly adds some new methods.

The aiodns release note says added support for getaddrinfo, and its getaddrinfo that is failing! That seems significant.

I've spent all day on this, and need to take a break now, I can investigate pycares or c-areas another day if the change to getaddrinfo can be ruled out.

@ColinRobbins
Author
ColinRobbins commented 16 hours ago
As per my report above, it also worked for me with a simple ncat server running on xxx.local as well.
However, when connecting to gateway-0403-XXXX-7195.local the getaddrinfo is failing under aiodns 3.2, but OK on 3.1.

Its not clear to me what getaddrinfo does. Is it actually contacting the host - is the response in someway unexpected?

@bdraco
Member
bdraco commented 16 hours ago
I'm also able to resolve an mdns .local name using the aiohttp AsyncResolver. Example below:

homeassistant:/config/issue_10111# python3 aiohttp_dns_test.py 
[{'hostname': 'drivewaygate.local.', 'host': '192.168.212.44', 'port': 80, 'family': <AddressFamily.AF_INET: 2>, 'proto': 0, 'flags': <AddressInfo.AI_NUMERICHOST|AI_NUMERICSERV: 1028>}]
homeassistant:/config/issue_10111# cat aiohttp_dns_test.py 
from aiohttp import AsyncResolver
import asyncio


async def aiohttp_test():
    resolver = AsyncResolver()
    host = await resolver.resolve('drivewaygate.local.', 80)
    print(host)

asyncio.run(aiohttp_test())

homeassistant:/config/issue_10111# python3 aiohttp_dns_test.py 
[{'hostname': 'drivewaygate.local.', 'host': '192.168.212.44', 'port': 80, 'family': <AddressFamily.AF_INET: 2>, 'proto': 0, 'flags': <AddressInfo.AI_NUMERICHOST|AI_NUMERICSERV: 1028>}]
homeassistant:/config/issue_10111# pip3 show aiohttp
Name: aiohttp
Version: 3.11.10.dev0
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache-2.0
Location: /usr/local/lib/python3.13/site-packages
Requires: aiohappyeyeballs, aiosignal, attrs, frozenlist, multidict, propcache, yarl
Required-by: accuweather, adax, Adax-local, adguardhome, advantage-air, AEMET-OpenData, afsapi, agent-py, aio-geojson-geonetnz-quakes, aio-geojson-geonetnz-volcano, aio-geojson-nsw-rfs-incidents, aio-geojson-usgs-earthquakes, aio-georss-client, aio_geojson_client, aio_geojson_generic_client, aioairq, aioairzone, aioairzone-cloud, aioambient, aioaquacell, aioaseko, aioautomower, aioazuredevops, aiobotocore, aiocomelit, aiodukeenergy, aioeafm, aioeagle, aioecowitt, aioelectricitymaps, aioemonitor, aioflo, aiogithubapi, aioguardian, aioharmony, aiohasupervisor, aiohttp-cors, aiohttp-fast-zlib, aiohttp-retry, aiohttp-sse-client2, aiohttp_socks, aiohue, aiolivisi, aiolookin, aiolyric, aiomealie, aiomodernforms, aiomusiccast, aionanoleaf, aionotion, aiooncue, aioopenexchangerates, aiopegelonline, aiopurpleair, aiopvapi, aiopvpc, aiopyarr, aioqsw, aiorecollect, aioresponses, aioridwell, aioruckus, aioshelly, aioskybell, aioslimproto, aiosolaredge, AIOSomecomfort, aiosteamist, aiostreammagic, aioswitcher, aiosyncthing, aiotankerkoenig, aiotedee, aiotractive, aiounifi, aiovodafone, aiowaqi, aiowatttime, aiowithings, aioymaps, airgradient, airly, airthings_cloud, alpha-vantage, anova-wifi, apple_weatherkit, apsystems-ez1, asmog, async_upnp_client, asyncarve, asyncpysupla, asyncsleepiq, autarco, authcaptureproxy, ayla_iot_unofficial, blebox_uniapi, blinkpy, bond-async, bring-api, brunt, connect_box, crownstone-cloud, crownstone-sse, deebot-client, demetriek, devialet, dio-chacon-wifi-api, directv, DoorBirdPy, easyenergy, electrickiwi-api, elgato, eliqonline, elvia, emulated-roku, energyzero, enturclient, env_canada, epson-projector, esphome-dashboard-api, eternalegypt, evohome-async, faadelays, firebase-messaging, fivem-api, foobot-async, forecast-solar, freebox-api, freenub, fyta_cli, gcal_sync, geniushub-client, geocachingapi, gios, go2rtc-client, goalzero, google_nest_sdm, google_photos_library_api, goslide-api, gotailwind, greeneye-monitor, gridnet, hass-nabucasa, here_routing, here_transit, hko, hole, homeassistant, homematicip, imgw_pib, incomfort-client, intellifire4py, jsonrpc-async, jsonrpc-websocket, kaiterra-async-client, knocki, konnected, lacrosse_view, laundrify-aio, linear-garage-door, london-tube-status, loqedAPI, mac-vendor-lookup, matrix-nio, mbddns, meater-python, microBeesPy, mill-local, millheater, moehlenhoff-alpha2, monarchmoney, monzopy, motioneye-client, mozart_api, music_assistant_client, mutesync, mypermobil, myuplink, nettigo_air_monitor, nexia, nextcord, nextdns, nice-go, nyt_games, odp-amsterdam, omnilogic, open-garage, open-meteo, opensensemap-api, openwebifpy, opower, ourgroceries, ovoenergy, p1monitor, panasonic-viera, peco, plexauth, plexwebsocket, plugwise, poolsense, ProgettiHWSW, pvo, py-aosmith, py-ccm15, py-melissa-climate, py-synologydsm-api, py_nightscout, pyaftership, pyairnow, pyairvisual, pyasuswrt, pyatag, pyatmo, pyatv, pyaussiebb, pyblu, pybravia, pycfdns, pyControl4, pydaikin, pydeconz, pydelijn, pydrawise, pydroid-ipcam, pyebox, pyeconet, pyefergy, pyElectra, pyemoncms, pyeverlights, pyevilgenius, pyfido, pyforked-daapd, PyFronius, pygti, pyhaversion, pyhiveapi, pyinsteon, pyintesishome, pyipma, pyipp, pyiqvia, pyiskra, pyisy, pykaleidescape, pykmtronic, pykodi, pykoplenti, pylaunches, pylibrespot-java, pylitterbot, PyLoadAPI, pymelcloud, PyMetEireann, PyMetno, PyNINA, pynordpool, pynws, pyoctoprintapi, pyopenuv, pyopenweathermap, pyosoenergyapi, pyoverkiz, pypalazzetti, pypoint, pyprosegur, pyps4-2ndscreen, pyrisco, pyrituals, pyrympro, pysabnzbd, pysensibo, pyseventeentrack, pyskyqhub, pysma, pysmartthings, pysmlight, pyspcwebgw, pysqueezebox, pysuezV2, pyswitchbee, PySwitchbot, PySyncThru, pytautulli, python-awair, python-bsblan, python-family-hub-local, python-homeassistant-analytics, python-homewizard-energy, python-izone, python-juicenet, python-kasa, python-matter-server, python-mystrom, python-otbr-api, python-roborock, python-smarttub, python-songpal, python-technove, python_linkplay, python_opendata_transport, python_opensky, pyTibber, pytile, pytomorrowio, pytouchlinesl, pytraccar, pytrafikverket, pyuptimerobot, pyversasense, pyvizio, pyvolumio, pywmspro, pyxeoma, radios, regenmaschine, renault-api, reolink_aio, ring-doorbell, rokuecp, romy, rtsp-to-webrtc, sense_energy, sharkiq, simplefin4py, simplepush, simplisafe-python, sisyphus_control, slackclient, smart_meter_texas, snitun, solarlog_cli, solax, sonos-websocket, spotifyaio, starline, stookwijzer, subarulink, surepy, switchbot_api, systembridgeconnector, tailscale, tesla-wall-connector, tesla_fleet_api, tesla_powerwall, teslajsonpy, teslemetry_stream, thinqconnect, toonapi, tp-connected, tplink_omada_client, ttls, twentemilieu, twitchAPI, uasiren, uiprotect, unifi-discovery, universal-silabs-flasher, vallox_websocket_api, vehicle, volkszaehler, volvooncall, vulcan-api, weatherflow4py, webmin-xmlrpc, whirlpool_sixth_sense, wled, xbox-webapi, xiaomi-ble, yalexs, yolink-api, youtubeaio, zamg, zigpy, zwave-js-server-python
homeassistant:/config/issue_10111# pip3 show aiodns
Name: aiodns
Version: 3.2.0
Summary: Simple DNS resolver for asyncio
Home-page: https://github.com/saghul/aiodns
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /usr/local/lib/python3.13/site-packages
Requires: pycares
Required-by: aiodiscover, easyenergy, forecast-solar, homeassistant, pyjvcprojector, radios, slixmpp, vulcan-api
@bdraco
Member
bdraco commented 16 hours ago • 
Its not clear to me what getaddrinfo does. Is it actually contacting the host - is the response in someway unexpected?

getaddrinfo is a standard library call for resolving a name

https://en.wikipedia.org/wiki/Getaddrinfo
https://man7.org/linux/man-pages/man3/getaddrinfo.3.html
https://stackoverflow.com/questions/64238919/understanding-getaddrinfo-function-in-c

In this case, this is the underlying method

https://c-ares.org/docs/ares_getaddrinfo.html

@bdraco
Member
bdraco commented 16 hours ago • 
home-assistant/core#132318 (comment)

Traceback (most recent call last):
  File "/home/colin/github/Home-Assistant-Core/venv/lib/python3.12/site-packages/aiohttp/resolver.py", line 103, in resolve
    resp = await self._resolver.getaddrinfo(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
aiodns.error.DNSError: (11, 'Could not contact DNS servers')
This implies the problem is a local configuration issue as the configured DNS servers are refusing connections

@bdraco
Member
bdraco commented 16 hours ago
https://github.com/c-ares/c-ares/blob/a8c091750d4db206ead3555b9a18e238525d7ca8/src/lib/ares_strerror.c#L56

@ColinRobbins
Author
ColinRobbins commented 16 hours ago
I agree this implies a DNS issue, but it works fine for aiodns 3.1, and the issue is occuring on two different local computers, including one running an unmodified HA OS. And there are several people reporting the issue.

@bdraco
Member
bdraco commented 15 hours ago • 
If you have aiodns 3.1.0 installed, aiohttp won't use the AsyncResolver and will fallback to the ThreadedResolver because it will be missing the getaddrinfo method it needs in aiodns.

aiohttp/aiohttp/resolver.py

Line 13 in 703ce61

 aiodns_default = hasattr(aiodns.DNSResolver, "getaddrinfo") 
But that doesn't explain why ThreadedResolver would work differently since it also uses getaddrinfo

aiohttp/aiohttp/resolver.py

Line 36 in 703ce61

 infos = await self._loop.getaddrinfo( 
@ColinRobbins
Author
ColinRobbins commented 15 hours ago
Is this an mDNS issue?
The address that is failing is not configured in DNS anywhere, it a server that has announced itself to the network.

@bdraco
Member
bdraco commented 15 hours ago
Is this an mDNS issue?

Almost certainly as .local domains are almost always going to be mDNS

@ColinRobbins
Author
ColinRobbins commented 15 hours ago • 
I don’t really understand the details of mDNS, but doesn’t it reply on getting a response from server - which will be need to be asynchronous in some way.
Is it possible the new async, as opposed to threaded approach is not getting these responses, so throwing an error.

@bdraco
Member
bdraco commented 14 hours ago
I don’t really understand the details of mDNS, but doesn’t it reply on getting a response from server - which will be need to be asynchronous in some way. Is it possible the new async, as opposed to threaded approach is not getting these responses, so throwing an error.

Can you try the examples I posted above in your Home Assistant container?

#10111 (comment)
#10111 (comment)

Replace drivewaygate.local. with the name you expect to be able to resolve

@ColinRobbins


Add a comment
Comment
There was a problem saving your comment. Please try again.

Here is the output of the test...

```
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ cat dns_test.py 
import aiodns
import asyncio
import socket

ADDRESS = "gateway-0403-XXXX-7195.local"

async def main():
    resolver = aiodns.DNSResolver()
    try:
        result = await resolver.query(ADDRESS, "A")
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.gethostbyname(ADDRESS, socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        result = await resolver.getaddrinfo(ADDRESS, 0, socket.AF_INET)
        print(result)
    except aiodns.error.DNSError as e:
        print(e)


asyncio.run(main())
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ python3.12 dns_test.py 
(11, 'Could not contact DNS servers')
(11, 'Could not contact DNS servers')
(11, 'Could not contact DNS servers')
(venv) colin@mars:~/github/Home-Assistant-Core/venv$ pip3 show aiodns
Name: aiodns
Version: 3.2.0
Summary: Simple DNS resolver for asyncio
Home-page: https://github.com/saghul/aiodns
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /home/colin/github/Home-Assistant-Core/venv/lib/python3.12/site-packages
Requires: pycares
Required-by: homeassistant
```

I downgraded to aiodns 3.1.0 and get the same (except test 3 give a different error), which is very odd. 
 
Remember, contributions to this repository should follow its contributing guidelines, security policy, and code of conduct.
Assignees
No one assigned
Labels
bug
needs-info
reproducer: missing
Projects
None yet
Milestone
No milestone
Development
No branches or pull requests

Loading
3 participants
@bdraco
@Dreamsorcerer
@ColinRobbins
Footer
© 2024 GitHub, Inc.
Footer navigation
Terms
Privacy
Security
Status
Docs
Contact
Manage cookies
Do not share my personal information



```

I downgraded to aiodns 3.1.0 and get the same (except test 3 give a different error), which is very odd.

**Comment by bdraco** [comment]
I would suggest using `strace -f -s 4096 python3 dns_test.py` to determine which syscalls are being made

**Comment by ColinRobbins** [comment]
Here is some code that shows the problem consicely.
getaddrinfo in aiodns fails, whereas the asyncio version works...

```
(venv) $ cat dns_test.py 
import aiodns
import asyncio
import socket

ADDRESS = "gateway-0403-XXXX-7195.local"

async def main():
    resolver = aiodns.DNSResolver()
    try:
        result = await resolver.getaddrinfo(ADDRESS,
            port=8443, 
            type=socket.SOCK_STREAM,
            family=socket.AF_INET,
            flags=socket.AI_ADDRCONFIG,
        )
        print(result)
    except aiodns.error.DNSError as e:
        print(e)
    try:
        loop = asyncio.get_running_loop()
        result = await loop.getaddrinfo(ADDRESS, 
            port=8443, 
            type=socket.SOCK_STREAM,
            family=socket.AF_INET,
            flags=socket.AI_ADDRCONFIG,
        )
        print(result)
    except aiodns.error.DNSError as e:
        print(e)


asyncio.run(main())
(venv) $ python3.12 dns_test.py
(11, 'Could not contact DNS servers')
[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('192.168.xx.xx', 8443))]
(venv) $ pip show aiodns
Name: aiodns
Version: 3.2.0
Summary: Simple DNS resolver for asyncio
Home-page: https://github.com/saghul/aiodns
Author: Saúl Ibarra Corretgé
Author-email: s@saghul.net
License: MIT
Location: /home/colin/github/HomeAssistant/venv/lib/python3.12/site-packages
Requires: pycares
Required-by: 
```

**Comment by bdraco** [comment]
```
9220  socket(AF_INET, SOCK_DGRAM, IPPROTO_IP) = 6
9220  fcntl(6, F_GETFL)                 = 0x2 (flags O_RDWR)
9220  fcntl(6, F_SETFL, O_RDWR|O_NONBLOCK) = 0
9220  fcntl(6, F_SETFD, FD_CLOEXEC)     = 0
9220  connect(6, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, 16) = 0
9220  epoll_ctl(3, EPOLL_CTL_ADD, 6, {events=EPOLLIN, data={u32=6, u64=130601365536774}}) = 0
9220  sendto(6, "p\245\1\0\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 46, MSG_NOSIGNAL, NULL, 0) = 46
9220  epoll_wait(3, [{events=EPOLLIN, data={u32=6, u64=130601365536774}}], 2, 1000) = 1
9220  recvfrom(6, "p\245\205\202\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 4097, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [16]) = 46
9220  sendto(6, "p\245\1\0\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 46, MSG_NOSIGNAL, NULL, 0) = 46
9220  recvfrom(6, 0x7fff4716e7c0, 4097, 0, 0x7fff4716e7a0, [16]) = -1 EAGAIN (Resource temporarily unavailable)
9220  epoll_wait(3, [{events=EPOLLIN, data={u32=6, u64=130601365536774}}], 2, 999) = 1
9220  recvfrom(6, "p\245\205\202\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 4097, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [16]) = 46
9220  sendto(6, "p\245\1\0\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 46, MSG_NOSIGNAL, NULL, 0) = 46
9220  recvfrom(6, 0x7fff4716e7c0, 4097, 0, 0x7fff4716e7a0, [16]) = -1 EAGAIN (Resource temporarily unavailable)
9220  epoll_wait(3, [{events=EPOLLIN, data={u32=6, u64=130601365536774}}], 2, 999) = 1
9220  recvfrom(6, "p\245\205\202\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 4097, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [16]) = 46
9220  sendto(6, "p\245\1\0\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 46, MSG_NOSIGNAL, NULL, 0) = 46
9220  recvfrom(6, 0x7fff4716e7c0, 4097, 0, 0x7fff4716e7a0, [16]) = -1 EAGAIN (Resource temporarily unavailable)
9220  epoll_wait(3, [{events=EPOLLIN, data={u32=6, u64=130601365536774}}], 2, 998) = 1
9220  recvfrom(6, "p\245\205\202\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1", 4097, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [16]) = 46
9220  epoll_ctl(3, EPOLL_CTL_DEL, 6, 0x7fff4716e374) = 0
9220  close(6)                          = 0
9220  epoll_wait(3, [], 1, 0)           = 0
9220  write(1, "(11, 'Could not contact DNS servers')\n", 38) = 38
```

**Comment by bdraco** [comment]
```
% python3
Python 3.12.7 (main, Oct  1 2024, 02:05:46) [Clang 16.0.0 (clang-1600.0.26.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dns.message
>>> dns.message.from_wire(b"p\245\205\202\0\1\0\0\0\0\0\0\26gateway-0403-5459-7195\5local\0\0\1\0\1").to_text()
'id 28837\nopcode QUERY\nrcode SERVFAIL\nflags QR AA RD RA\n;QUESTION\ngateway-0403-5459-7195.local. IN A\n;ANSWER\n;AUTHORITY\n;ADDITIONAL'
>>> print('id 28837\nopcode QUERY\nrcode SERVFAIL\nflags QR AA RD RA\n;QUESTION\ngateway-0403-5459-7195.local. IN A\n;ANSWER\n;AUTHORITY\n;ADDITIONAL')
id 28837
opcode QUERY
rcode SERVFAIL
flags QR AA RD RA
;QUESTION
gateway-0403-5459-7195.local. IN A
;ANSWER
;AUTHORITY
;ADDITIONAL
>>> 
```


It looks like everything with aiodns is working as expected, and the dns server is responding with SERVFAIL

**Comment by ColinRobbins** [comment]
So, how come the call to the asyncio version of getattrinfo works?

**Comment by bdraco** [comment]
The threaded resolver will fallback to using avahi because the system has `/lib/x86_64-linux-gnu/libnss_mdns4_minimal.so.2` configured.

```
9223  connect(6, {sa_family=AF_UNIX, sun_path="/run/avahi-daemon/socket"}, 110) = 0
9223  fcntl(6, F_GETFL)                 = 0x2 (flags O_RDWR)
9223  fstat(6, {st_mode=S_IFSOCK|0777, st_size=0, ...}) = 0
9223  write(6, "RESOLVE-HOSTNAME-IPV4 gateway-0403-5459-7195.local\n", 51) = 51
9223  read(6, "+ 3 1 gateway-0403-5459-7195.local 192.168.10.187\n", 4096) = 50
9223  close(6)                          = 0
9223  sendto(5, "\0", 1, 0, NULL, 0)    = 1
```

**Comment by ColinRobbins** [comment]
Sorry, not sure what that means?

Does that mean the functions are not equivalent?  So should it be rolled back?

What do we need to do to make it work?

If the issue is to the system configuration, do we need to get HA OS updated?

**Comment by bdraco** [comment]
> Sorry, not sure what that means?
> 
> Does that mean the functions are not equivalent? So should it be rolled back?
> 

It means the system has a custom resolver plugin (https://github.com/avahi/nss-mdns) that `c-ares` (https://c-ares.org/#supported-rfcs-and-proposals) and be extension `aiodns` doesn't know about or support.

> What do we need to do to make it work?
> 
> If the issue is to the system configuration, do we need to get HA OS updated?

The plugin is a `nss-mdns` plugin is for `glibc` and not `musl` (Alpine Linux) so it doesn't look like you are not running the test script inside the HAOS container so the result may be a red herring

**Comment by bdraco** [comment]
Make sure you are in the `homeassistant` container when you are running the test as the system resolver will be different in the container.

`docker exec -it homeassistant /bin/bash`

**Comment by ColinRobbins** [comment]
I get exactly the same issue in HAOS - that's why I am spending all this time debugging it!

As HAOS does not expose python or ptrace I had to run on my HA test system, which run in a venv.

**Comment by bdraco** [comment]
> I get exactly the same issue in HAOS - that's why I am spending all this time debugging it!

> As HAOS does not expose python or ptrace I had to run on my HA test system, which run in a venv.

You can install `strace` in the container with `apk add strace`

**Comment by ColinRobbins** [comment]
The problem I am seeing is repeatable on the HA OS - that why I am debugging.   Everything worked until 2024.12, when aiodns 3.2.0 was added.
This is being seen by multiple people.

I only have access to the HA OS via the terminal and SSH plugin.   This does not expose python, trace or docker.
So I ran the test in my HA development environment, on a different machine using Python Venv.

```

       ▄██▄           _   _
     ▄██████▄        | | | | ___  _ __ ___   ___
   ▄████▀▀████▄      | |_| |/ _ \| '_ ` _ \ / _ \
 ▄█████    █████▄    |  _  | (_) | | | | | |  __/
▄██████▄  ▄██████▄   |_| |_|\___/|_| |_| |_|\___|          _
████████  ██▀  ▀██      / \   ___ ___(_)___| |_ __ _ _ __ | |_
███▀▀███  ██   ▄██     / _ \ / __/ __| / __| __/ _` | '_ \| __|
██    ██  ▀ ▄█████    / ___ \\__ \__ \ \__ \ || (_| | | | | |_
███▄▄ ▀█  ▄███████   /_/   \_\___/___/_|___/\__\__,_|_| |_|\__|
▀█████▄   ███████▀

Welcome to the Home Assistant command line.

System information
  IPv4 addresses for end0:  192.168..xx.xxx24
  IPv6 addresses for end0:  xxxxxxxxx
  IPv4 addresses for wlan0:

  OS Version:               Home Assistant OS 14.0
  Home Assistant Core:      2024.11.2

  Home Assistant URL:       http://homeassistant.local:8123
  Observer URL:             http://homeassistant.local:4357
[core-ssh ~]$
[core-ssh ~]$
[core-ssh ~]$
[core-ssh ~]$
[core-ssh ~]$ docker exec -it homeassistant /bin/bash
```

I am a HA novice at this level.   I just seeing something that working was in 2024.11 not working in 2024.12.
On my test system Venv - the tests / traces are what I am seeing - and highlight the issue.

So, if some kind of OS fix is needed - a similar OS fix is needed in HA OS.

**Comment by bdraco** [comment]
As this issue is opened with `aiodns`, I'm providing support from the perspective of `aiodns` and previously `aiohttp`.  This doesn't seem like a problem with `aiodns` and helping with HAOS is out of scope for support here, please continue in the Home Assistant issue instead.

**Comment by ColinRobbins** [comment]
This is issue exposed in HA OS  following the upgrade for 2024.11 to 2024.12.

The code worked perfectly in HA 2024.11, using aiodns 3.1.0
HA 2024.12 is using aiodns 3.2.0 and it breaks.

The code fragments show how it breaks.   So the issue is in the 3.1.0 to 3.2.0 update not being backwards compatible.

**Comment by iMicknl** [comment]
I will try to provide a trace and see if I can pinpoint the issue, but I will first need to revive my Somfy hub.

@ColinRobbins the reason a trace from HA OS is needed is because this issue might be related to the networking stack, which is very OS (system packages etc.) dependent. Even though caused by an update in this package, the bug/culprit might be in the networking layer.

**Comment by ColinRobbins** [comment]
If someone can talk me through how to run the test in the HA OS, I can do it - but I only have access via the terminal add-on, which does not expose docker, python or trace.

**Comment by richardfudge-ovo** [comment]
I use "Terminal & SSH" version 9.15.0 and HAOS and I can get a shell inside the HomeAssistant container from there by running:
`docker exec -it homeassistant bash`

Then from there I can add the strace command by running
'apk add strace'

Then I can run 'strace python3 dns_test.py' (the results are not helpful as this is a rolled back version of HA)

**Comment by ColinRobbins** [comment]
So I figured out how to log into HA OS as a developer.
I then rerun the tests with aiodns 3.2.0 and the test_dns.py works perfectly.

So back to the drawing board, looks like its not a DNS issue after all.    
That seem to be a redherring (as you suggested) from my developer environment.

Sorry for wasting your time, but also thank you for you support - I have learnt a huge amount.

Now to re-research the original issue.
